### PR TITLE
Added docs for user migration to Asgardeo

### PR DESCRIPTION
**Pull request by himeshsiriwardana**
## Purpose

Related issue: https://github.com/wso2-enterprise/asgardeo-product/issues/24278

> SS For password migration docs:

index.md

image

Migrate user accounts
screencapture-localhost-8000-asgardeo-docs-guides-users-migrate-users-to-asgardeo-migrate-users-2024-07-19-15_15_58

Migrate user passwords
screencapture-localhost-8000-asgardeo-docs-guides-users-migrate-users-to-asgardeo-migrate-passwords-2024-07-19-15_16_23

Adaptive auth script
screencapture-localhost-8000-asgardeo-docs-guides-authentication-conditional-auth-on-demand-silent-password-migration-template-2024-07-19-15_17_16

### CODE REVIEW

File: en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+    === "Classic Editor"
+
+        To configure the login flow using the classic editor:
+
+        1. Click **Start with default configuration** or add the **Username & Password** authenticator to the first step of the flow.
+        
+        2. Turn on Conditional Authentication by switching on the toggle.
+
+            ![Make the polling endpoint open]({{base_path}}/assets/img/guides/users/migrate-users/migrate-users-classic.png)
```
Comment (ThaminduDilshan) [inline en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md:181]:
```suggestion
            ![Configure with classic editor]({{base_path}}/assets/img/guides/users/migrate-users/migrate-users-classic.png)
```

File: en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+        To configure the login flow using the visual editor:
+
+        1. Switch to the Visual Editor tab.
+
+        2. If **Username & Password** authentication is not set, go to **Predefined flows** > **Basic Login Flows** > **Add Basic Login** and select it to add to the login flow.
+
+        3. Turn on Conditional Authentication by switching on the toggle.
+
+             ![Make the polling endpoint open]({{base_path}}/assets/img/guides/users/migrate-users/migrate-users-visual.png){: width="500"}
```
Comment (ThaminduDilshan) [inline en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md:195]:
```suggestion
             ![Configure with visual editor]({{base_path}}/assets/img/guides/users/migrate-users/migrate-users-visual.png){: width="500"}
```

File: en/includes/guides/authentication/conditional-auth/on-demand-silent-password-migration-template.md [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+    - If the entered credentials match, the user's password is already migrated. Hence, the `onSuccess` function will be called and the user will be authenticated.
+    
+    - If the user's credentials don't match, either the credentials are incorrect or the password may not have be migrated yet. Hence, the `onFail` callback function is called and the script continues.
+
+2. If the `onFail` function is called, the script will next try to locate a unique user in the system using the `resolveAndInitUser` function. 
+
+    !!! note "Adjust script for alternate login identifiers"
+
+        If your organization does not use alternate login identifiers, uncomment the following lines in the `resolveAndInitUser` function.
```
Comment (ThaminduDilshan) [inline en/includes/guides/authentication/conditional-auth/on-demand-silent-password-migration-template.md:344]:
```suggestion
        If your organization does not use alternate login identifiers, comment the following lines in the `resolveAndInitUser` function.
```

We should call `resolveMultiAttributeLoginIdentifier` function only when alternative login is enabled

File: en/includes/guides/authentication/conditional-auth/on-demand-silent-password-migration-template.md [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+
+    - The `connectionMetadata` object holds the start authentication endpoint URL, Choreo application credentials and the Asgardeo token endpoint.
+    - The script first calls the `callChoreo()` function along with the `connectionMetaData` and invokes the start authentication endpoint.
+    - If the API call is successful, the `onSuccess()` callback function is called which in turn calls the `prompt()` function.
+    - The `prompt()` function continuously polls the Choreo polling endpoint and redirects the user to a waiting page until the external authentication completes.
+    - Once the authentication is complete, the `onSuccess()` callback function of the `prompt()` function calls the `updatePasswordAndReAuthenticate()` function.
+
+5. The `updatePasswordAndReAuthenticate()` function is responsible for checking the status of the authentication and taking necessary actions as explained below.
+
+    - The `connectionMetadata` object hold the authentication status endpoint URL, Choreo application credentials and the Asgardeo token endpoint
```
Comment (ThaminduDilshan) [inline en/includes/guides/authentication/conditional-auth/on-demand-silent-password-migration-template.md:368]:
```suggestion
    - The `connectionMetadata` object hold the authentication status endpoint URL, Choreo application credentials and the Asgardeo token endpoint.
```

File: en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+1. The user logs in to the application.
+
+2. The application redirects the user to the Asgardeo login screen.
+
+3. The user enters the credentials used for the account in the legacy IdP.
+
+4. If Asgardeo has already migrated the user's password, the user is logged in. If not, the user is redirected to a waiting page.
+
+5. Asgardeo makes an authentication request to the legacy IdP with the user provided credentials.
```
[the remark this answers is not in the document]
Comment (himeshsiriwardana) [reply en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md]: Fixed with https://github.com/wso2/docs-is/pull/4669/commits/d0b7130af4b2f0857f356c6dbdcbce192e8484c2

File: en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+!!! note
+    You may find a sample Ballerina authentication service [here](https://github.com/wso2/samples-is/tree/master/user-migration-samples/asgardeo/external-authentication-service){target="_blank"}.
+
+1. Create a new Ballerina package. Learn how to do so in the [Ballerina documentation](https://ballerina.io/learn/get-started/){target="_blank"}.
+
+2. Write a [RESTful service](https://ballerina.io/learn/write-a-restful-api-with-ballerina/){target="_blank"} that communicates with the legacy authentication system to perform user authentication. Your service should expose the following REST API endpoints.
+
+    !!! tip
+        
+        External authentication might take a significant processing time depending on the legacy IdP and network delays. Therefore it is recommended to perform the external authentication asynchronously without holding an active connection until the external authentication process completes.
```
[the remark this answers is not in the document]
Comment (himeshsiriwardana) [reply en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md]: Reworded.
"External authentication may require additional processing time"

Fixed with https://github.com/wso2/docs-is/commit/d0b7130af4b2f0857f356c6dbdcbce192e8484c2

File: en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+!!! note
+
+    To learn more about applications in Choreo, refer to the [Choreo documentation](https://wso2.com/choreo/docs/consuming-services/manage-application/#create-an-application){target="_blank"}.
+
+
+### Step 4: Configure an attribute in Asgardeo to hold the migration status
+
+Asgardeo should have means to check whether a user's password is migrated or not. This can be achieved by creating a user attribute in Asgardeo to hold the migration status. To do so,
+
+1. On the Asgardeo Console, go to **User Attributes & Stores** > **Attributes** and click on **Attirubtes**.
```
[the remark this answers is not in the document]
Comment (himeshsiriwardana) [reply en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md]: Fixed with https://github.com/wso2/docs-is/commit/d0b7130af4b2f0857f356c6dbdcbce192e8484c2

File: en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md (hunk 1 above)
Comment (himeshsiriwardana) [reply en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md]: Fixed with https://github.com/wso2/docs-is/commit/d0b7130af4b2f0857f356c6dbdcbce192e8484c2

File: en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md (hunk 2 above)
Comment (himeshsiriwardana) [reply en/asgardeo/docs/guides/users/migrate-users-to-asgardeo/migrate-passwords.md]: Fixed with https://github.com/wso2/docs-is/commit/d0b7130af4b2f0857f356c6dbdcbce192e8484c2

File: en/includes/guides/authentication/conditional-auth/on-demand-silent-password-migration-template.md (hunk 4 above)
Comment (himeshsiriwardana) [reply en/includes/guides/authentication/conditional-auth/on-demand-silent-password-migration-template.md:368]: Fixed with https://github.com/wso2/docs-is/commit/d0b7130af4b2f0857f356c6dbdcbce192e8484c2

**Review by ThaminduDilshan — DISMISSED** [review]

**Comment by himeshsiriwardana** [comment]
The latest commits brought the adaptive script methods into documentation

prompt() method

image

resolveMultiAttributeLoginIdentifier() and updateUserPassword() methods

image

File: en/includes/references/conditional-auth/api-reference.md [hunk 8]
```diff
@@ -886,7 +1030,79 @@ When using httpGet or httpPost functions in Asgardeo adaptive authentication scr
...
+- **Parameters**
+
+    <table>
+      <tbody>
+        <tr>
+          <td><code>loginIdentifier</code></td>
+          <td>User provided login identifier.</td>
+        </tr>
+        <tr>
+          <td><code>tenantDomain</code></td>
```
Comment (ThaminduDilshan) [inline en/includes/references/conditional-auth/api-reference.md:1051]:
```suggestion
          <td><code>organization</code></td>
```

**Review by ThaminduDilshan — DISMISSED** [review]

**Review by ThaminduDilshan — APPROVED** [review]

**Pull request merged**